src/main/java/org/truffleruby/language/LazyRubyNode.java

```diff
@@ -63,11 +63,13 @@ public RubyNode resolve() {
                 Log.LOGGER.info(() -> "lazy translating " + RubyLanguage.fileLine(getParent().getEncapsulatingSourceSection()) + " in " + getRootNode());
             }
 
-            resolved = resolver.get();
-            transferFlagsTo(resolved);
+            final RubyNode result = resolver.get();
+            transferFlagsTo(result);
 
-            replace(resolved, "lazy node resolved");
-            return resolved;
+            // publish
```

Comment:
By writing to this volatile variable, I am trying to achieve [Safe Initialization](https://shipilev.net/blog/2014/safe-public-construction/#_safe_initialization).
While the code going through the monitor is all fine, once the LazyRubyNode is replaced it's just a normal and unsynchronized field read from the parent node to access the new node.
I want the reader to observe all effects that happened before the assignment.
By publication semantics, if the object was never published to another thread before and there is a StoreStore barrier (here modeled by the volatile write) between assignments to its field and the publishing write, I believe we get these guarantees.
I could use an explicit barrier if you think that's better.
The proper, clean solution would be to always keep the LazyRubyNode around (no `replace()`) and always go through a volatile read. We can also treat the read constant after the first read, through `@CompilationFinal/@Child`.